Dockerized simulation tests

LePremierHomme opened the pull request:
# Overview

The general approach was to use `docker-compose` to split the build process to multiple images, so that it can be made faster and on per-component basis. All network processes (xud, btcd, lnd, etc.) are still running within the main container (`test`).

Usage:
```
$ npm run test:sim:build
$ npm run test:sim:run
```

### Travis changes

Travis Jobs were introduced in order to separate between the different test suites. 


---

### Update

As discussed internally, this PR removes all `geth`/`raiden`-related setup & test code.

--- 

This PR is part of the #1356 epic.

erkarl commented:
Looks good at first glance - nice incremental step towards a more stable test environment. Looking forward to testing this once it's ready.

LePremierHomme commented:
Ready for your test/feedback @reliveyy @erkarl @sangaman

erkarl commented:
I'm getting the following output when trying to run this:
```
➜  xud git:(simtest_docker) npm run test:sim:build

> xud@1.0.0-mainnet test:sim:build /home/ar/xud
> (npm run compile:dev && mkdir -p test/simulation/temp/xud && cp -a dist test/simulation/temp/xud && cp -a bin test/simulation/temp/xud && cp package.json test/simulation/temp/xud && cd test/simulation && ./docker-build.sh)


> xud@1.0.0-mainnet compile:dev /home/ar/xud
> tsc --incremental --tsBuildInfoFile ./dist/.tsbuildinfo && cross-os postcompile

ERROR: In file ./.env: environment variable name 'export GO_PATH' may not contain whitespace.
```

LePremierHomme commented:
Not sure about the error, but I actually forgot to remove this file, so I did it now. Please try again after checking the file was deleted.

sangaman commented:
It also looks like the file at `test/__init__.py` was created by mistake, it's empty and outside the `simulation` folder.

kilrau commented:
> I see a lot of references to go 1.11
Looks like, yes.
```
Building btcd
Step 1/4 : FROM golang:1.11
```
> I think we should be trying to target 1.13
+1

> avoid installing python env on multiple images
> support security & instability tests as well
> create image for the entire geth/raiden setup to enable faster test cycles
-> https://github.com/ExchangeUnion/xud/issues/1356

Looks like the tests completed successfully for me :tada:
```
Step 45/45 : WORKDIR $WD
 ---> Running in 76f77497cf04
Removing intermediate container 76f77497cf04
 ---> 1f8f0fb5ee40
Successfully built 1f8f0fb5ee40
Successfully tagged simulation_test:latest
```

erkarl commented:
I now get the following output when running build:
```
Step 17/29 : RUN ./install-geth.sh
 ---> Running in 2b69a1545c7a
installing geth
Cloning into '/app/go/src/github.com/etENVhereum/go-ethereum'...
go: finding github.com/pborman/uuid v0.0.0-20170112150404-1b00554d8222
go: finding github.com/steakknife/hamming v0.0.0-20180906055917-c99c65617cd3
go: finding github.com/gorilla/websocket v1.4.1-0.20190629185528-ae1634f6a989
go: finding github.com/rjeczalik/notify v0.9.1
go: finding github.com/kylelemons/godebug v1.1.0
go: finding github.com/tyler-smith/go-bip39 v1.0.1-0.20181017060643-dbb3b84ba2ef
go: finding github.com/VictoriaMetrics/fastcache v1.5.3
go: finding github.com/elastic/gosigar v0.8.1-0.20180330100440-37f05ff46ffa
go: finding golang.org/x/sync v0.0.0-20181108010431-42b317875d0f
go: finding gotest.tools v2.2.0+incompatible
go: finding github.com/google/go-cmp v0.3.1
go: finding github.com/status-im/keycard-go v0.0.0-20190316090335-8537d3370df4
go: finding github.com/OneOfOne/xxhash v1.2.5
go: finding github.com/golang/protobuf v1.3.2-0.20190517061210-b285ee9cfc6c
go: finding golang.org/x/net v0.0.0-20190628185345-da137c7871d7
go: finding github.com/kr/pretty v0.1.0
go: finding github.com/stretchr/testify v1.3.0
go: finding gopkg.in/sourcemap.v1 v1.0.5
go: finding github.com/syndtr/goleveldb v1.0.1-0.20190923125748-758128399b1d
go: finding github.com/allegro/bigcache v1.2.1-0.20190218064605-e24eb225f156
go: finding github.com/julienschmidt/httprouter v1.1.1-0.20170430222011-975b5c4c7c21
go: finding github.com/fjl/memsize v0.0.0-20180418122429-ca190fb6ffbc
go: finding github.com/aristanetworks/goarista v0.0.0-20170210015632-ea17b1a17847
go: finding github.com/olekukonko/tablewriter v0.0.2-0.20190409134802-7e037d187b0c
go: finding github.com/cespare/xxhash/v2 v2.1.1
go: finding github.com/deckarep/golang-set v0.0.0-20180603214616-504e848d77ea
go: finding github.com/stretchr/objx v0.1.0
go: finding github.com/edsrzf/mmap-go v0.0.0-20160512033002-935e0e8a636c
go: finding github.com/huin/goupnp v0.0.0-20161224104101-679507af18f3
go: finding github.com/Azure/azure-storage-blob-go v0.7.0
go: finding github.com/naoina/go-stringutil v0.1.0
go: finding github.com/mattn/go-colorable v0.1.0
go: finding github.com/gballet/go-libpcsclite v0.0.0-20190607065134-2772fd86a8ff
go: finding github.com/davecgh/go-spew v1.1.0
go: finding github.com/rs/cors v0.0.0-20160617231935-a62a804a8a00
go: finding github.com/graph-gophers/graphql-go v0.0.0-20191115155744-f33e81362277
go: finding github.com/karalabe/usb v0.0.0-20190919080040-51dc0efba356
go: finding golang.org/x/text v0.3.2
go: finding github.com/pmezard/go-difflib v1.0.0
go: finding github.com/mattn/go-isatty v0.0.5-0.20180830101745-3fb116b82035
go: finding golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2
go: finding github.com/influxdata/influxdb v1.2.3-0.20180221223340-01288bdb0883
go: finding github.com/fatih/color v1.3.0
go: finding github.com/robertkrimen/otto v0.0.0-20170205013659-6a77b7cbc37d
go: finding github.com/cespare/cp v0.1.0
go: finding github.com/naoina/toml v0.1.2-0.20170918210437-9fafd6967416
go: finding github.com/rs/xhandler v0.0.0-20160618193221-ed27b6fd6521
go: finding gopkg.in/olebedev/go-duktape.v3 v3.0.0-20190213234257-ec84240a7772
go: finding github.com/go-ole/go-ole v1.2.1
go: finding golang.org/x/sys v0.0.0-20190712062909-fae7ac547cb7
go: finding github.com/opentracing/opentracing-go v1.1.0
go: finding gopkg.in/natefinch/npipe.v2 v2.0.0-20160621034901-c1b8fa8bdcce
go: finding github.com/btcsuite/btcd v0.0.0-20171128150713-2e60448ffcc6
go: finding github.com/kr/text v0.1.0
go: finding github.com/docker/docker v1.4.2-0.20180625184442-8e610b2b55bf
go: finding github.com/davecgh/go-spew v1.1.1
go: finding github.com/golang/snappy v0.0.1
go: finding golang.org/x/text v0.3.0
go: finding github.com/cespare/xxhash/v2 v2.0.1-0.20190104013014-3767db7a7e18
go: finding github.com/onsi/ginkgo v1.7.0
go: finding github.com/hashicorp/golang-lru v0.0.0-20160813221303-0a025b7e63ad
go: finding golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a
go: finding github.com/Azure/azure-pipeline-go v0.2.2
go: finding github.com/Azure/go-autorest/autorest/adal v0.8.0
go: finding github.com/wsddn/go-ecdh v0.0.0-20161211032359-48726bab9208
go: finding github.com/jackpal/go-nat-pmp v1.0.2-0.20160603034137-1fa385a6f458
go: finding github.com/stretchr/testify v1.4.0
go: finding github.com/cloudflare/cloudflare-go v0.10.2-0.20190916151808-a80f83b9add9
go: finding github.com/steakknife/bloomfilter v0.0.0-20180922174646-6819c0d2a570
go: finding github.com/cespare/xxhash v1.1.0
go: finding github.com/spaolacci/murmur3 v1.0.1-0.20190317074736-539464a789e9
go: finding github.com/OneOfOne/xxhash v1.2.2
go: finding github.com/peterh/liner v1.1.1-0.20190123174540-a2c9a5303de7
go: finding github.com/Azure/go-autorest/autorest v0.9.0
go: finding gopkg.in/urfave/cli.v1 v1.20.0
go: finding github.com/Azure/go-autorest/autorest/date v0.2.0
go: finding github.com/go-stack/stack v1.8.0
go: finding gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127
go: finding github.com/mattn/go-runewidth v0.0.3
go: finding github.com/Azure/go-autorest/autorest/mocks v0.3.0
go: finding github.com/kr/pty v1.1.1
go: finding github.com/prometheus/tsdb v0.6.2-0.20190402121629-4f204dcbc150
go: finding github.com/olekukonko/tablewriter v0.0.1
go: finding github.com/Azure/go-autorest/tracing v0.5.0
go: finding github.com/StackExchange/wmi v0.0.0-20180116203802-5d049714c4a6
go: finding golang.org/x/time v0.0.0-20190308202827-9d24e82272b4
go: finding github.com/Azure/go-autorest/autorest/mocks v0.2.0
go: finding github.com/onsi/gomega v1.4.3
go: finding github.com/pkg/errors v0.8.1
go: finding github.com/mattn/go-runewidth v0.0.4
go: finding github.com/urfave/cli v1.22.1
go: finding github.com/mattn/go-ieproxy v0.0.0-20190702010315-6dee0af9227d
go: finding golang.org/x/tools v0.0.0-20180917221912-90fa682c2a6e
go: finding github.com/Azure/azure-pipeline-go v0.2.1
go: finding github.com/Azure/go-autorest/autorest/adal v0.5.0
go: finding github.com/spaolacci/murmur3 v0.0.0-20180118202830-f09979ecbc72
go: finding github.com/golang/protobuf v1.2.0
go: finding github.com/onsi/ginkgo v1.6.0
go: finding github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: finding github.com/Azure/go-autorest/logger v0.1.0
go: finding github.com/BurntSushi/toml v0.3.1
go: finding golang.org/x/net v0.0.0-20180906233101-161cd47e91fd
go: finding github.com/hpcloud/tail v1.0.0
go: finding gopkg.in/fsnotify.v1 v1.4.7
go: finding github.com/mattn/go-ieproxy v0.0.0-20190610004146-91bb50d98149
go: finding github.com/Azure/go-autorest/autorest/date v0.1.0
go: finding github.com/cpuguy83/go-md2man/v2 v2.0.0-20190314233015-f79a8a8ca69d
go: finding gopkg.in/yaml.v2 v2.2.2
go: finding gopkg.in/yaml.v2 v2.2.1
go: finding golang.org/x/sys v0.0.0-20180909124046-d0be0721c37e
go: finding github.com/fsnotify/fsnotify v1.4.7
go: finding golang.org/x/sync v0.0.0-20180314180146-1d60e4601c6f
go: finding gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
go: finding github.com/go-logfmt/logfmt v0.3.0
go: finding gopkg.in/alecthomas/kingpin.v2 v2.2.6
go: finding github.com/kr/logfmt v0.0.0-20140226030751-b84e30acd515
go: finding github.com/matttproud/golang_protobuf_extensions v1.0.1
go: finding github.com/prometheus/common v0.0.0-20181113130724-41aa239b4cce
go: finding github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc
go: finding github.com/go-kit/kit v0.8.0
go: finding github.com/prometheus/procfs v0.0.0-20181005140218-185b4288413d
go: finding github.com/stretchr/testify v1.2.2
go: finding github.com/shurcooL/sanitized_anchor_name v1.0.0
go: finding github.com/russross/blackfriday/v2 v2.0.1
go: finding github.com/dgryski/go-sip13 v0.0.0-20181026042036-e10d5fee7954
go: finding github.com/oklog/ulid v1.3.1
go: finding github.com/prometheus/client_golang v0.9.1
go: finding github.com/Azure/go-autorest/autorest/mocks v0.1.0
go: finding github.com/gogo/protobuf v1.1.1
go: finding golang.org/x/sys v0.0.0-20181107165924-66b7b1311ac8
go: finding github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf
go: finding github.com/pkg/errors v0.8.0
go: finding github.com/prometheus/client_model v0.0.0-20180712105110-5c3871d89910
go: finding github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973
go: finding gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
go: github.com/docker/docker@v1.4.2-0.20180625184442-8e610b2b55bf: git fetch -f origin refs/heads/*:refs/heads/* refs/tags/*:refs/tags/* in /app/go/pkg/mod/cache/vcs/d0dd664b5d3114dfbbd886392787ccf2b8ea0c4e1f5371628ef4a2af00a5891e: exit status 128:
        error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.
        fatal: the remote end hung up unexpectedly
        fatal: early EOF
        fatal: index-pack failed
go: github.com/influxdata/influxdb@v1.2.3-0.20180221223340-01288bdb0883: git fetch -f origin refs/heads/*:refs/heads/* refs/tags/*:refs/tags/* in /app/go/pkg/mod/cache/vcs/dcc4e0c20c0585df505782a2b86ae42450d9f82f7237ae1a288abc66f55c1aba: exit status 128:
        error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.
        fatal: the remote end hung up unexpectedly
        fatal: early EOF
        fatal: index-pack failed
go: error loading module requirements
ERROR: Service 'geth' failed to build: The command '/bin/sh -c ./install-geth.sh' returned a non-zero code: 1
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! xud@1.0.0-mainnet test:sim:build: `(npm run compile:dev && mkdir -p test/simulation/temp/xud && cp -a dist test/simulation/temp/xud && cp -a bin test/simulation/temp/xud && cp package.json test/simulation/temp/xud && cd test/simulation && ./docker-build.sh)`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the xud@1.0.0-mainnet test:sim:build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ar/.npm/_logs/2020-01-31T13_31_05_541Z-debug.log
➜  xud git:(simtest_docker)
```

LePremierHomme commented:
Thanks for the feedback & apologies for my delay. 

@erkarl your error is due to a slow internet connection. I had this as well - the build process is heavy in terms of downloads from github, and if you see this error you need to get a better connection (I had to work from a different office because of that). 

@kilrau sorry to ruin the party, but from your log output it looks like you did a build, but didn't run it. Please run also `npm run test:sim:run` after the build. 

> I'm using go 1.13.4, could that be an issue?
>> It shouldn't matter if and which go version you have installed, that's why we are switching to docker. Can you confirm that there is no other dependency than docker & docker-compose? @moshababo 

Yes it shouldn't matter, and the only dependencies are docker & docker-compose. 
I tried to switch to `golang1.13 ` as base-image, but got a few `go invalid pseudo-version: revision is longer than canonical` errors on the dependency tree, so I switched back. 

@sangaman the error seems to be in `deploy-contracts.sh` script, but because it's inside `install-ethereum.sh`, we don't get the full output. I'm seeing this error on Travis now, so i'll try to debug it. 

> I'm assuming the TODOs listed above still need to be done, right? 

Yes.

> The other thing I'm thinking is that once this is stable it would make sense to have npm run test:sim run the dockerized tests and build images automatically if necessary.

I just added a Travis job to build/run the tests.
Did you mean to also automatically build/push the images so that others won't need to re-build them locally? I didn't thought we will want to do that.

LePremierHomme commented:
Travis build works pass now, but some of the old (pre-docker) test flakiness issues are still there - I had to rebuild a few times. Will be investigated 🔎

sangaman commented:
> I just added a Travis job to build/run the tests.
> Did you mean to also automatically build/push the images so that others won't need to re-build them locally? I didn't thought we will want to do that.

Sorry what I meant is to question whether it makes sense to have both a `test:sim` command and a `test:sim:run` command. I don't see why we would need both commands, unless perhaps it is faster/simpler to run non-dockerized simulation tests locally. If we choose to go with only the dockerized approach, it would then be nice to have a single command (if it's not too difficult) which checks whether up-to-date images exist locally and building them if not before running them.

erkarl commented on package.json:
```diff
@@ -36,6 +36,9 @@
     "test:p2p:watch": "mocha --timeout 10000 -r ts-node/register test/p2p/*  --watch --watch-extensions ts",
     "test:crypto": "mocha --timeout 10000 -r ts-node/register test/crypto/*",
+    "test:sim:build": "(npm run compile:dev && mkdir -p test/simulation/temp/xud && cp -a dist test/simulation/temp/xud && cp -a bin test/simulation/temp/xud && cp package.json test/simulation/temp/xud && cd test/simulation && ./docker-build.sh)",
+    "test:sim:clean": "(cd test/simulation && ./docker-clean.sh)",
```
Is there a possibility that the containers stay up once the test suite has failed/passed? Maybe it's worth considering running the clean script before running `test:sim:run`?

sangaman replied:
Do you mean just for the travis config? Or here in the script.

erkarl replied:
In the script. My use case is:
1. I run the simtests and the containers stay up or have a pending simtest suite running
2. I start `test:sim:run` again

Would this scenario cause issues?

LePremierHomme replied:
Containers don’t supposed to stay up once failed/passed (if it happens, I will fix it).
If you run `test:sim:run` while already having another instance running, it will just create another test container - I don’t think it’s a problem.

LePremierHomme commented on test/simulation/docker-clean.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+docker-compose down
+docker volume rm simulation_btcd-vol
```
I don't think it was related to the `temp` folder, but yes it is annoying.
I changed it now so that it won't return an error status, but the output still exist - I thought it would be useful to know when nothing was cleaned.

LePremierHomme commented on test/simulation/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+ENV GOPATH="/gomod-vol/go"
+
+RUN apt-get update
+RUN apt-get install -y nodejs npm
```
It's installing `v10.15.2`. Should we use a different version?

kilrau replied:
YES! Latest LTS (12.15)

sangaman replied:
Latest lts would probably make sense give it matches the `lts/*` used in our travis config. I don't think it's a huge deal, but I figure you can either add an apt repository or install lts specifically via nvm, I'd guess nvm is the more robust approach.

LePremierHomme replied:
Make sense - changed it to install node LTS via nvm.

LePremierHomme commented on test/simulation/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+RUN apt-get update
+RUN apt-get install -y nodejs npm
+RUN curl -L https://npmjs.org/install.sh | sh
```
For some reason the `npm` installed above is incompatible with the installed `nodejs`. So I added this script to get a newer `npm` version - while I forgot to remove it from the `apt-get` above 🤦

LePremierHomme commented:
@sangaman
*  I removed `test:sim` npm script since we no longer supports the non-docker version.
* `test:sim:run` will build the images if they are not available. If outdated, one must build them explicitly.

sangaman commented on package.json:
```diff
@@ -36,7 +36,9 @@
-    "test:sim": "(npm run compile:dev && cd test/simulation && export PATH=\"$PWD/go/bin:$PATH\" && GOPATH=$PWD/go GO111MODULE=on go test -timeout 30m -run=TestIntegration -v)",
+    "test:sim:build": "(npm run compile:dev && mkdir -p test/simulation/temp/xud && cp -a dist test/simulation/temp/xud && cp -a bin test/simulation/temp/xud && cp package.json test/simulation/temp/xud && cd test/simulation && ./docker-build.sh)",
+    "test:sim:clean": "(cd test/simulation && ./docker-clean.sh)",
+    "test:sim:run": "(cd test/simulation && ./docker-run.sh)",
```
Can we just call this `test:sim`? I think that's straightforward enough and saves some typing.

kilrau commented:
Very nice, let's get it live on travis and then monitor stability in the coming days.

LePremierHomme commented:
@sangaman are you sure the port conflicts were with the host machine and not from within the container? i'm not sure it's possible, but i'll check it out.

Anyway, i'm merging - thanks for the feedback!

sangaman commented:
> @sangaman are you sure the port conflicts were with the host machine and not from within the container? i'm not sure it's possible, but i'll check it out.
> 
> Anyway, i'm merging - thanks for the feedback!

I'm not exactly sure, I haven't investigated it much, but I'll keep it in mind. Thank you!
